sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java

```diff
@@ -0,0 +1,338 @@
+     * @param asyncHttpDecodedResponse the asynchronous HTTP response to the original HTTP request
+     * @param methodParser the SwaggerMethodParser that the request originates from
+     * @param returnType the type of value that will be returned
+     * @param context Additional context that is passed through the Http pipeline during the service call.
+     * @return the deserialized result
+     */
+    private Object handleRestReturnType(final Mono<HttpResponseDecoder.HttpDecodedResponse> asyncHttpDecodedResponse,
+                                        final SwaggerMethodParser methodParser,
+                                        final Type returnType,
+                                        final Context context,
+                                        final RequestOptions options,
+                                        final EnumSet<ErrorOptions> errorOptionsSet) {
+        final Mono<HttpResponseDecoder.HttpDecodedResponse> asyncExpectedResponse =
+            ensureExpectedStatus(asyncHttpDecodedResponse, methodParser, options, errorOptionsSet)
+                .doOnEach(AsyncRestProxy::endTracingSpan)
+                .contextWrite(reactor.util.context.Context.of("TRACING_CONTEXT", context));
+
+        final Object result;
+        if (TypeUtil.isTypeOrSubTypeOf(returnType, Mono.class)) {
+            final Type monoTypeParam = TypeUtil.getTypeArgument(returnType);
+            if (TypeUtil.isTypeOrSubTypeOf(monoTypeParam, Void.class)) {
+                // ProxyMethod ReturnType: Mono<Void>
+                result = asyncExpectedResponse.then();
+            } else {
+                // ProxyMethod ReturnType: Mono<? extends RestResponseBase<?, ?>>
+                result = asyncExpectedResponse.flatMap(response ->
+                    handleRestResponseReturnType(response, methodParser, monoTypeParam));
+            }
+        } else if (FluxUtil.isFluxByteBuffer(returnType)) {
+            // ProxyMethod ReturnType: Flux<ByteBuffer>
+            result = asyncExpectedResponse.flatMapMany(ar -> ar.getSourceResponse().getBody());
+        } else if (TypeUtil.isTypeOrSubTypeOf(returnType, void.class) || TypeUtil.isTypeOrSubTypeOf(returnType,
+            Void.class)) {
+            // ProxyMethod ReturnType: Void
+            asyncExpectedResponse.block();
+            result = null;
+        } else {
+            // ProxyMethod ReturnType: T where T != async (Mono, Flux) or sync Void
+            // Block the deserialization until a value T is received
+            result = asyncExpectedResponse
+                .flatMap(httpResponse -> handleRestResponseReturnType(httpResponse, methodParser, returnType))
+                .block();
+        }
+        return result;
+    }
+
+    // This handles each onX for the response mono.
+    // The signal indicates the status and contains the metadata we need to end the tracing span.
+    private static void endTracingSpan(Signal<HttpResponseDecoder.HttpDecodedResponse> signal) {
```

Comment:
I think that:
- `endTracingSpan` from sync could move to base and be made `protected`
- Here we could simply extract inputs needed for sync version from `signal` and call into sync version (after moving to base).